chronos/asyncmacro2.nim

```diff
@@ -353,8 +264,32 @@ proc asyncSingleProc(prc: NimNode): NimNode {.compileTime.} =
   #if prcName == "recvLineInto":
   #  echo(toStrLit(result))
 
-proc await*[T](x: T) =
-  {.error: "Await only available within {.async.}".}
+template await*[T](f: Future[T]): auto =
+  when declared(chronosInternalRetFuture):
+    when not declared(chronosInternalTmpFuture):
+      var chronosInternalTmpFuture: FutureBase
+    chronosInternalTmpFuture = f
+    chronosInternalRetFuture.child = chronosInternalTmpFuture
+    yield chronosInternalTmpFuture
+    chronosInternalTmpFuture.internalCheckComplete()
+    chronosInternalRetFuture.child = nil
+    cast[type(f)](chronosInternalTmpFuture).internalRead()
+  else:
+    static:
+      assert(false, "Await only available within {.async.}")
```

Comment:
Why not raise an explicit exception instead of having it as a side effect of a failing assertion (which, BTW, doesn't work in release mode)?

Reply:
It's in the `static:` block. Compile time only.

Reply:
You were right. `assert` doesn't work in release :). Fixed in subsequent pr.

Reply:
Static assert() behaviour is very inconsistent:
- in Nim 0.19.6 it's disabled with -d:release
- in 0.20.2 it works with -d:release, but not with -d:danger

I'd say it's best to avoid it all the time.